bonfhir/bonfhir1_archive
# Make `$expand` `filter` parameter optional

## What is this about

This change makes the `filter` param optional n `ExpandOperationParameters` as specified by its [cardinality in FHIR][0]. Along the way, documentation was added from the online FHIR spec for each of the `ExpandOperationParameters` and `GraphOperationParameters`.

## Issue ticket numbers

Fixes #37 by making `filter` param optional

As mentioned in #37, we could move the param types and their type guards to `core`. I'm looking for confirmation about [introducing the remaining `$expand` params](https://github.com/bonfhir/bonfhir/issues/37#issuecomment-1403983188) before adding more changes.

## How can this be tested?

Using the `FhirRestfulClientAdapter` for Medplum, call `execute` to fetch a given ValueSet without providing any `filter` param. It should succeed and return the ValueSet unfiltered.

## Known limitations/edge cases

I've opted for a slightly stricter check for required strings, checking that they contain some non-whitespace characters. The rationale here is that the FHIR servers are expected to trim string inputs removing leading and trailing whitespace and empty strings are invalid wherever strings are required. Thus, checking this upfront increases the likelihood of requests succeeding and avoids possibly unnecessary requests. I can relax this to guarding against empty string if there's a risk of dropping valid requests here.

[0]: https://www.hl7.org/fhir/valueset-operation-expand.html
